pool.go

```diff
@@ -171,23 +171,59 @@ func (p *connPool) new() (*conn, error) {
 		)
 		return nil, err
 	}
+
 	cn, err := p.dial()
 	if err != nil {
 		p.lastDialErr = err
+		return nil, err
 	}
-	return cn, err
+
+	if err := p.init(cn); err != nil {
+		p.Remove(cn)
```

Comment:
Yes. `Remove` also opens new connection, so we can have an infinitive loop here.